### ISSUE TITLE
Implement approvals in IGSDB

### ISSUE THREAD

**Issue by christoph-maurer** [building-envelope-data/api#303]
Approvals are a cryptographic signature which states that the data sets fulfills certain requirements. They will be used by software applications to filter suitable high-quality data sets.

Currently we think about approvals for NFRC and AERC and possibly for LBNL research level and generic records.

**Comment by christoph-maurer** [comment]
ISE will prepare a proposal.

**Comment by danielmcquillen** [comment]
@christoph-maurer Where does the proposal for this feature stand? Is it related to the issue that mentions hashes?

**Comment by RDmitchell** [comment]
@christoph-maurer -- in order to understand what this issue entails, we would need a detailed proposal/spec.

**Comment by christoph-maurer** [comment]
@RDmitchell @danielmcquillen The approvals are defined in https://github.com/building-envelope-data/api/blob/develop/apis/database.graphql#L1507-L1794 . @StephenCzarnecki and @simon-wacker had the discussion https://github.com/building-envelope-data/api/issues/187 which may be helpful. We have set the priority of this issue to low because the [synchronization of the uuids of components and institutions](https://github.com/LBNL-ETA/igsdb-webapp/pull/229) has a much higher priority for the software companies.

**Comment by simon-wacker** [comment]
@danielmcquillen Currently the specification of approvals in https://github.com/building-envelope-data/api/blob/develop/apis/database.graphql#L1507-L1794 needs some polishing. I started working on it but then got held up by https://github.com/LBNL-ETA/igsdb-webapp/pull/229 I'll continue with the polishing / fine-tuning / updating the specs as soon as possible.

**Comment by simon-wacker** [comment]
@danielmcquillen @RDmitchell At least for reference: I polished the specification for approvals in the API documentation: https://github.com/building-envelope-data/api/blob/dc15b3c891040ba193bf932a4e9803218c67f835/apis/database.graphql#L1681-L2202 To support it minimally, the IGSDB would just need to have a one-to-many relationship from physical properties, optical properties, and detail files to one or three new tables that hold(s) the [8 scalar values of data approvals](https://github.com/building-envelope-data/api/blob/dc15b3c891040ba193bf932a4e9803218c67f835/apis/database.graphql#L2154-L2182). To support it in a more complex fashion, it could validate approvals before they are added to the database and/or support the creation of such approvals for end users.

**Comment by christoph-maurer** [comment]
@simonwacker has implemented ResponseApprovals for IGSDB with the graphql field `approval`. DataApprovals are not yet implemented. Therefore, the graphql field `approvals` returns an empty array.
```graphql
{
  allOpticalData {
    edges {
      node {
        approvals {
          query
          signature
        }
        approval {
          query
          signature
        }
      }
    }
  }
}
```

**Comment by simonwacker** [comment]
No, I haven't implemented response approvals. I have only added the GraphQL fields needed to make the GraphQL schema conform to the standard https://github.com/building-envelope-data/api/blob/develop/apis/database.graphql in the superficial way of returning some data for all non-optional response approval fields. In particular, the signature does not contain a signature (as far as I remember it is an empty string). Data approvals on the other hand are conform to the standard in every way, in the sense that every data approval that is in the IGSDB is returned and for every data approval, every field has a proper value. However, there is no way to add a data approval to the IGSDB, so it does not have any data approvals, thus satisfying the standard trivially. The same trivial support does not work for response approvals because for each data entry, its response approval needs to be non-null when asked for (so I cannot just return `null` which would be the pendant to the empty array I return for data approvals). The reason why serving response approvals even with dummy-data is better then not serving that at all is that then at least the GraphQL schema conforms to the standard and therefore queries which are valid for https://github.com/building-envelope-data/api/blob/develop/apis/database.graphql are also valid for the IGSDB. This is the minimal requirement that allows the metabase to communicate seamlessly with the IGSDB (even when asking for response approvals). Because response approvals hadn't been supported before by the IGSDB, [the metabase did not (and still does not but could) ask for response approvals in the first place (which obviously is a bad hack to work around an issue with one product-data database)](https://github.com/building-envelope-data/metabase/blob/f78a5d52efb51bb6f7e09831ca94d81314da3050/backend/src/GraphQl/DataX/Data.cs#L37).

**Comment by christoph-maurer** [comment]
Thank you @simonwacker for the correction and the detailed explanation!
